docs/development/design/build-images.rst

```diff
@@ -0,0 +1,274 @@
+* use a ``base`` image with the dependencies that don't change frequently (OS and base requirements)
+* ``base`` image naming is tied to the OS version (e.g. Ubuntu LTS)
+* allow us to update a Python version without affecting the ``base`` image
+* reduce size on builder VM disks by sharing Docker image layers
+* allow users to specify extra dependencies (apt packages, node, rust, etc)
+* allow use "limited" custom images for users by sharing most layers
+* automatically build & push *all* images on commit
+* deprecate ``stable``, ``latest`` and ``testing``
+
+
+New build image structure
+-------------------------
+
+.. Taken from https://github.com/readthedocs/readthedocs-docker-images/blob/master/Dockerfile
+
+* ``ubuntu20-base``
+
+  * labels
+  * environment variables
+  * system dependencies
+  * install requirements
+  * LaTeX dependencies (for PDF generation)
+  * other languages version managers (``pyenv``, ``nodenv``, etc)
+  * UID and GID
+
+The following images all are based on ``ubuntu20-base``:
+
+* ``ubuntu20-py*``
+
+  * Python version installed via ``pyenv``
+  * default Python packages (pinned versions)
+    * pip
+    * setuptools
+    * virtualenv
+  * labels
+
+* ``ubuntu20-conda*``
+
+  * same as ``-py*`` versions
+  * ``mamba`` executable
+
+Note that all these images only need to run ``pyenv install ${PYTHON_VERSION}``
+to install a specific Python/Conda version.
+
+.. Build all these images with Docker
+
+   docker build -t readthedocs/build:ubuntu20-base -f Dockerfile.base .
+   docker build -t readthedocs/build:ubuntu20-py39 -f Dockerfile.py39 .
+   docker build -t readthedocs/build:ubuntu20-conda47 -f Dockerfile.conda47 .
+
+   Check the shared space between images
+   docker system df --verbose | grep -E 'SHARED SIZE|readthedocs'
+
+   Initial Dockerfile.* as example for this are pushed in this PR
+   https://github.com/readthedocs/readthedocs-docker-images/pull/166
+
+
+Specifying extra users' dependencies
+------------------------------------
+
+Different users may have different requirements. We were already requested to install
+``swig``, ``imagemagick``, ``libmysqlclient-dev``, ``lmod``, ``rust``, ``poppler-utils``, etc.
+
+People with specific dependencies will be able to install them as APT packages or as extras
+using ``.readthedocs.yaml`` config file. Example:
+
+.. code:: yaml
+
+   build:
+     image: ubuntu20-py39
+     apt:
+       - swig
+       - imagemagick
+     extras:
+       - node==14.16
+       - rust==1.46.0
+
+
+.. note:: Idea for implementation
+
+   Once this config file is parsed, the builder builds a Docker image on-demand with a command similar to:
+
+   .. code:: bash
+
+      docker build \
+        --tag ${BUILD_ID} \
+        --file Dockerfile.custom \
+        --build-arg RTD_IMAGE=ubuntu20-py39
+        --build-arg RTD_NODE_VERSION=14.16.0 \
+        --build-arg RTD_RUST_VERSION=1.46.0 \
+        --build-arg RTD_APT_PACKAGES="swig imagemagick"
```

Comment:
I'm nor sure about building the docker image before using it, don't think there is an easy way to do this with our scaling sets

Reply:
Why are you referring to scale sets here?

The image will be built on the builder immediately before the starting the docs build process and after used for that particular build, it will be deleted:

* clone repository and parse config file
* `docker build ...`
* our build commands
* `docker rmi ${BUILD_ID}`